PolyEngine/Engine/Res/Shaders/lightAccumulationTexDebug.frag.glsl

```diff
@@ -0,0 +1,192 @@
+
+#ifndef MAX_DIRLIGHT_COUNT
+#define MAX_DIRLIGHT_COUNT 8
+#endif
+
+in VERTEX_OUT{
+	vec3 positionInWorld;
+	vec2 uv;
+	vec3 normalInModel;
+	mat3 tangentFromWorld;
+	vec3 viewPositionInTangent;
+	vec3 fragmentPositionInTangent;
+} fragment_in;
+
+
+struct DirectionalLight
+{
+	vec4 ColorIntensity;
+	vec4 Direction;
+};
+
+struct Light
+{
+	vec4 Position;
+	vec4 Color;
+	vec4 RangeIntensity;
+};
+
+struct VisibleIndex
+{
+	int Index;
+};
+
+layout(std430, binding = 0) readonly buffer LightBuffer {
+	Light data[];
+} bLightBuffer;
+
+layout(std430, binding = 1) readonly buffer VisibleLightIndicesBuffer {
+	VisibleIndex data[];
+} bVisibleLightIndicesBuffer;
+
+struct Material
+{
+	vec4 Emissive;
+	vec4 Albedo;
+	float Roughness;
+	float Metallic;
+	float OpacityMaskThreshold;
+};
+
+const uint MAX_NUM_LIGHTS = 1024;
+
+uniform sampler2D uBrdfLUT;
+uniform samplerCube uIrradianceMap;
+uniform samplerCube uPrefilterMap;
+
+uniform sampler2D uEmissiveMap;
+uniform sampler2D uAlbedoMap;
+uniform sampler2D uRoughnessMap;
+uniform sampler2D uMetallicMap;
+uniform sampler2D uNormalMap;
+uniform sampler2D uAmbientOcclusionMap;
+
+uniform float uTime;
+uniform Material uMaterial;
+
+uniform DirectionalLight uDirectionalLight[8];
+uniform int uDirectionalLightCount;
+
+uniform int uLightCount;
+uniform int uWorkGroupsX;
+uniform int uWorkGroupsY;
+
+layout(location = 0) out vec4 oColor;
+layout(location = 1) out vec4 oNormal;
+
+const float PI = 3.14159265359;
+
+float DistributionGGX(vec3 N, vec3 H, float roughness)
+{
+	float a = roughness*roughness;
+	float a2 = a*a;
+	float NdotH = max(dot(N, H), 0.0);
+	float NdotH2 = NdotH*NdotH;
+
+	float nom = a2;
+	float denom = (NdotH2 * (a2 - 1.0) + 1.0);
+	denom = PI * denom * denom;
+
+	return nom / denom;
+}
+
+float GeometrySchlickGGX(float NdotV, float roughness)
+{
+	float r = (roughness + 1.0);
+	float k = (r*r) / 8.0;
+
+	float nom = NdotV;
+	float denom = NdotV * (1.0 - k) + k;
+
+	return nom / denom;
+}
+
+float GeometrySmith(vec3 N, vec3 V, vec3 L, float roughness)
+{
+	float NdotV = max(dot(N, V), 0.0);
+	float NdotL = max(dot(N, L), 0.0);
+	float ggx2 = GeometrySchlickGGX(NdotV, roughness);
+	float ggx1 = GeometrySchlickGGX(NdotL, roughness);
+
+	return ggx1 * ggx2;
+}
+
+vec3 fresnelSchlick(float cosTheta, vec3 F0)
+{
+	return F0 + (1.0 - F0) * pow(1.0 - cosTheta, 5.0);
+}
+
+vec3 fresnelSchlickRoughness(float cosTheta, vec3 F0, float roughness)
+{
+	return F0 + (max(vec3(1.0 - roughness), F0) - F0) * pow(1.0 - cosTheta, 5.0);
+}
+
+void main()
+{
+	mat3 WorldFromTangent = inverse(fragment_in.tangentFromWorld);
+
+	// Debug texture input
+	// oColor.rgb = emissive.rgb; // missing
+	// oColor.rgb = albedo.rgb; // ok
+	// oColor.rgb = normal.rgb * 0.5 + 0.5; // ok
+	// oColor.rgb = (WorldFromTangent * normal.rgb) * 0.5 + 0.5;
+	// oColor.rgb = vec3(roughness); // ok
+	// oColor.rgb = vec3(metallic); // ok
+	// oColor.rgb = vec3(ao); // ok
+	// oColor.rgb = fragment_in.normalInModel * 0.5 + 0.5;
```

Comment:
Used for debugging correct texture inputs. Every single useful comparison. It would nice to create developer debug view from one of those - it sells and showcases PBR feature. Do we have appropriate config changed in runtime ex. r.ShowDebug.Roughness or r.ShowDebug.Metallic?